components/org.wso2.identity.apps.taglibs.layout.controller/src/main/java/org/wso2/identity/apps/taglibs/layout/controller/MainTagHandler.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.com).
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.identity.apps.taglibs.layout.controller;
+
+import org.wso2.identity.apps.taglibs.layout.controller.core.LocalTemplateEngineCacheWithTier;
+
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.tagext.TagSupport;
+
+/**
+ * Tag handler class for the "main" tag
+ */
+public class MainTagHandler extends TagSupport {
+
+    private static final long serialVersionUID = -7314788638114472678L;
+    private String layoutName;
+    private String layoutFileRelativePath;
+    private Map<String, Object> data = new HashMap<String, Object>();
+    private Boolean devMode = false;
+    private String testLayoutFileRelativePath = "";
+    private LocalTemplateEngineCacheWithTier engine;
+
+    public void setLayoutName(String layoutName) {
+
+        this.layoutName = layoutName;
+    }
+
+    public void setLayoutFileRelativePath(String layoutFileRelativePath) {
+
+        this.layoutFileRelativePath = layoutFileRelativePath;
+    }
+
+    public void setData(Map<String, Object> data) {
+
+        this.data = data;
+    }
+
+    public void setDevMode(Boolean devMode) {
+
+        this.devMode = devMode;
+    }
+
+    public void setTestLayoutFileRelativePath(String testLayoutFileRelativePath) {
+
+        this.testLayoutFileRelativePath = testLayoutFileRelativePath;
+    }
+
+    /**
+     * This method will execute when the starting point of the "main" tag is reached
+     *
+     * @return int -> SKIP_BODY, EVAL_BODY_INCLUDE, EVAL_BODY_BUFFERED
+     * @throws JspException
+     */
+    public int doStartTag() throws JspException {
+
+        engine = new LocalTemplateEngineCacheWithTier();
+        try {
+            engine.execute(
+                    layoutName,
+                    pageContext.getServletContext().getResource(layoutFileRelativePath),
```

Comment:
Where is the `pageContext` defined and initialized? Can `getServletContext()` returns **null**

Reply:
It can't be null. In a tag life cycle before the doStartTag execution, setPageContext method will be executed. You can refer to this for more details. (https://docs.oracle.com/javaee/7/api/javax/servlet/jsp/PageContext.html)